bluemira/equilibria/opt_constraints.py

```diff
@@ -20,312 +20,253 @@
-in derivative based algorithms, such as those utilising gradient descent.
+Equilibrium optimisation constraint classes
 """  # noqa (W505)
 
 from abc import ABC, abstractmethod
+from copy import deepcopy
 from typing import List, Union
 
 import numpy as np
 
+from bluemira.equilibria.opt_constraint_funcs import (
+    Ax_b_constraint,
+    L2_norm_constraint,
+    coil_field_constraints,
+    coil_force_constraints,
+)
 from bluemira.equilibria.plotting import ConstraintPlotter
 from bluemira.geometry._deprecated_loop import Loop
 from bluemira.utilities.opt_problems import OptimisationConstraint
 from bluemira.utilities.tools import abs_rel_difference, is_num
 
 
-def objective_constraint(constraint, vector, grad, objective_function, maximum_fom=1.0):
+def _get_dummy_equilibrium(equilibrium, I_not_dI):
     """
-    Constraint function to constrain the maximum value of an NLOpt objective
-    function provided
+    Get a dummy equilibrium for current optimisation where the background response is
+    solely due to the plasma and passive coils.
 
-    Parameters
-    ----------
-    objective_function: callable
-        NLOpt objective function to use in constraint.
-    maximum_fom: float (default=1.0)
-        Value to constrain the objective function by during optimisation.
+    Notes
+    -----
+    When we do dI (current gradient) optimisation, the background vector includes the
+    contributions from the whole coilset (including active coils).
+
+    When we do I (current vector) optimisation, the background vector only includes
+    contributions from the passive coils (plasma).
     """
-    constraint[:] = objective_function(vector, grad) - maximum_fom
-    return constraint
+    if I_not_dI:
+        # TODO: Add passive coil contributions here
+        dummy = equilibrium.plasma_coil()
+        dummy.coilset = deepcopy(equilibrium.coilset)
+        equilibrium = dummy
+    return equilibrium
 
 
-def Ax_b_constraint(constraint, vector, grad, a_mat, b_vec, value, scale):  # noqa: N802
+class UpdateableConstraint(ABC):
     """
-    Constraint function of the form:
-        A.x - b < value
-
-    Parameters
-    ----------
-    constraint: np.ndarray
-        Constraint array (modified in place)
-    vector: np.ndarray
-        Variable vector
-    grad: np.ndarray
-        Constraint Jacobian (modified in place)
-    A_mat: np.ndarray
-        Response matrix
-    b_vec: np.ndarray
-        Target value vector
-    value: float
-        Target constraint value
-    scale: float
-        Current scale with which to calculate the constraints
+    Abstract base mixin class for an equilibrium optimisation constraint that is
+    updateable.
     """
-    constraint[:] = np.dot(a_mat, scale * vector) - b_vec - value
-    if grad.size > 0:
-        grad[:] = scale * a_mat
-    return constraint
+
+    @abstractmethod
+    def prepare(self, equilibrium, I_not_dI=False, fixed_coils=False):  # noqa :N803
```

Comment:
If this is intended as an ABC - `equilibrium` could be replaced by an arbitrary dependency (parametrisation?) that the constraint function is dependent on.

Updateable constraints with dependencies is likely outside the scope of this PR though - steps in the right direction nonetheless, I think.

I think there's still a motive for an ArrayConstraint more generally, as discussed in #1027.